02-ADTs/homework-02/hw-02.hs

```diff
@@ -0,0 +1,198 @@
+I 147 mice in the air, I’m afraid, but you might catch a bat, and
+E 2 148 #56k istereadeat lo d200ff] BOOTMEM
+It’s all quite confusing; clearly we need a program to sort through
+this mess. We’ve come up with some data types to capture the structure of this log file format:
+data MessageType = Info
+| Warning
+| Error Int
+deriving (Show, Eq)
+type TimeStamp = Int
+data LogMessage = LogMessage MessageType TimeStamp String
+| Unknown String
+deriving (Show, Eq)
+Note that LogMessage has two constructors: one to represent normallyformatted log messages, and one to represent anything else that does
+not fit the proper format.
+We’ve provided you with a module Log.hs containing these data
+type declarations, along with some other useful functions. Download
+Log.hs and put it in the same folder where you intend to put your
+homework assignment. Please name your homework assignment
+LogAnalysis.hs (or .lhs if you want to make it a literate Haskell
+document). The first few lines of LogAnalysis.hs should look like
+this:
+{-# OPTIONS_GHC -Wall #-}
+module LogAnalysis where
+import Log
+which sets up your file as a module named LogAnalysis, and imports the module from Log.hs so you can use the types and functions
+it provides.
+-}
+
+
+{-
+The first step is figuring out how to parse an individual
+message. Define a function
+parseMessage :: String -> LogMessage
+which parses an individual line from the log file. For example,
+parseMessage "E 2 562 help help"
+== LogMessage (Error 2) 562 "help help"
+cis 194: homework 2 3
+parseMessage "I 29 la la la"
+== LogMessage Info 29 "la la la"
+parseMessage "This is not in the right format"
+== Unknown "This is not in the right format"
+Once we can parse one log message, we can parse a whole log file.
+Define a function
+parse :: String -> [LogMessage]
+which parses an entire log file at once and returns its contents as a
+list of LogMessages.
+To test your function, use the testParse function provided in the
+Log module, giving it as arguments your parse function, the number
+of lines to parse, and the log file to parse from (which should also be
+in the same folder as your assignment). For example, after loading
+your assignment into GHCi, type something like this at the prompt:
+testParse parse 10 "error.log"
+Don’t reinvent the wheel! (That’s so last week.) Use Prelude functions to make your solution as concise, high-level, and functional as
+possible. For example, to convert a String like "562" into an Int, you
+can use the read function. Other functions which may (or may not)
+be useful to you include lines, words, unwords, take, drop, and (.).
+-}
+--Exercise 1
+parseMessage :: String -> LogMessage
+parseMessage str = case words str of
+  ("E" : mt : ts : msg) -> LogMessage (Error (read mt)) (read ts) (unwords msg)
+  ("I" : ts : msg)      -> LogMessage Info (read ts) (unwords msg)
+  ("W" : ts : msg)      -> LogMessage Warning (read ts) (unwords msg)
+  _                     -> Unknown str
```

Comment:
Using `read` is not the best choice here. It makes the assumption that the argument it takes is an `Int`. In this case it would fail at runtime if e.g. the 2nd field wasn't an int. User can test it with something like parseMessage "E invalid arg".

When building applications you'd want to avoid such runtime crashes. For this exercise, you should have a look at `readMaybe` which parses invalid cases into `Unknown`.

Reply:
Good point!

Reply:
Agreed. In general you want to avoid partial functions (like, head, tail, read etc.) most of the time.
Unfortunately `base` doesn't make it very clear *which* functions are partial, but you'll learn that with each new crash :) The safe alternative to use in this case is [`readMaybe`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Text-Read.html#v:readMaybe) and in case of failed parse falling back to Unknown constructor.

Reply:
I think @peironggg can come back to this and use `readMaybe` after reaching `applicatives`. Otherwise there will be quite a bit of nesting differentiate between `Nothing` and `Just _` values.

Reply:
I second @kwannoel; the exercise makes presumptions about which functions are to be used, and in this case (unfortunately), it seems like `read` was expected (given the type signature). 

Of course, there's always the option of pattern matching on the `Maybe` value, and throwing an `error` in case of a `Nothing`. I'll leave this up to @peironggg to decide. The important thing about this thread though, is that it emphasises the avoidance of partial functions in Haskell.